### DISCUSSION TITLE
help request: how to pick up username/password from consumer when accessing API?

### DISCUSSION THREAD

**Discussion by zhongpengrui** [apache/apisix#7510] [Q&A]
### Description

大佬好，我有这样一个需求：我使用了basic-auth插件作权限过滤，consumer中注册了用户，然后在路由中需要去访问其他的服务接口，需要将basic-auth插件中的username和password取出来，作为访问后面的路由的参数传递过去，即要对参数进行一次包装，请问这需要怎么去处理？

### Environment

- APISIX version (run `apisix version`):2.12.1
- Operating system (run `uname -a`):Linux bogon 3.10.0-1160.el7.x86_64 #1 SMP Mon Oct 19 16:18:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`):
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`):
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`):

**Comment by tzssangglass** [comment]
If you are talking about passing consumer name upstream, see: https://github.com/apache/apisix/issues/4078#issuecomment-822231019

and `basic-auth` plugin has `hide_credentials` attributes， set true to pass the authorization request headers to the Upstream.

**Comment by zhongpengrui** [comment]
您说得这种方式可以将信息写在headers中，但是不能动态变化。
我想实现在发送A请求之前先发送一个获取token的请求B，然后将token作为参数动态拼接在A请求之上，
您知道这个怎么处理吗？




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/apisix"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年7月20日(星期三) 上午9:52
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/apisix] help request: how to pick up username/password from consumer when accessing API? (Issue #7496)





 
If you are talking about passing consumer name upstream, see: #4078 (comment)
 
and basic-auth plugin has hide_credentials attributes， set true to pass the authorization request headers to the Upstream.

**Comment by tzssangglass** [comment]
> 您说得这种方式可以将信息写在headers中，但是不能动态变化。 我想实现在发送A请求之前先发送一个获取token的请求B，然后将token作为参数动态拼接在A请求之上， 您知道这个怎么处理吗？

It looks like you need to implement it yourself. 

Construct the http request during the authentication process, get the response and then modify the parameters of the current request.

**Reply by zhongpengrui** [reply]
您的意思是我需要自己开发一个插件，还是有其他什么方式通过编码实现，您这边有相关的示例或者处理过类似的问题吗？

**Reply by tzssangglass** [reply]
> 您的意思是我需要自己开发一个插件

yes

> 您这边有相关的示例或者处理过类似的问题吗？

There are, in the APISIX source code, many similar features that you can learn from. You can also take a look at: https://github.com/apache/apisix/blob/master/docs/en/latest/plugin-develop.md
